SRVCOM-1334: Add 1.16.0 serverless release notes (#33701)
openshift/openshift-docs: modules/serverless-rn-1-16-0.adoc

.. Apply the following network policy to allow traffic from the Prometheus namespace:
+
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-from-openshift-monitoring-ns
  namespace: knative-serving
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          name: "openshift-monitoring"
  podSelector: {}
  policyTypes:
  - Ingress
----

.. Modify and reapply the default Service Mesh control plane in the `istio-system` namespace, so that it includes the following spec:
+
[source,yaml]
----
spec:
  proxy:
    networking:
      trafficControl:
        inbound:
          excludedPorts:
          - 8444
----

* If you deploy Service Mesh CRs with the Istio ingress enabled, you may see the following warning in the `istio-ingressgateway` pod:
+
[source,terminal]
----
2021-05-02T12:56:17.700398Z warning envoy config [external/envoy/source/common/config/grpc_subscription_impl.cc:101] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Error adding/updating listener(s) 0.0.0.0_8081: duplicate listener 0.0.0.0_8081 found
----
+
Your Knative services may also not be accessible.
+

> You can use the following workaround to fix this issue by recreating the `knative-local-gateway` CR:

Reviewer: ```suggestion
You can use the following workaround to fix this issue by recreating the `knative-local-gateway` service:
```
The `knative-local-gateway` is not CR but kubernetes service.

> .. Delete the existing `knative-local-gateway` CR by deleting existing services in the `istio-system` namespace:

Reviewer: ```suggestion
.. Delete the existing `knative-local-gateway` service in the `istio-system` namespace:
```

+
[source,terminal]
----

> $ oc delete services -n istio-system

Reviewer: ```suggestion
$ oc delete services -n istio-system knative-local-gateway
```

Sorry, I missed `knative-local-gateway` the draft in the JIRA.

----

> .. Create and apply a new `knative-local-gateway` CR that contains the following YAML:

Reviewer: ```suggestion
.. Create and apply a new `knative-local-gateway` service that contains the following YAML:
```

[id="serverless-rn-1-16-0_{context}"]
= Release Notes for Red Hat {ServerlessProductName} 1.16.0

[id="new-features-1-16-0_{context}"]
== New features

* {ServerlessProductName} now uses Knative Serving 0.22.0.
* {ServerlessProductName} now uses Knative Eventing 0.22.0.
* {ServerlessProductName} now uses Kourier 0.22.0.
* {ServerlessProductName} now uses Knative `kn` CLI 0.22.0.
* {ServerlessProductName} now uses Knative Kafka 0.22.0.

> * The `kn func` CLI plugin now uses `func` 0.16.0.

Reviewer 2: ```suggestion
* The `kn func` CLI plug-in now uses `func` 0.16.0.
```

> 🚫  plugin noun, adjective
Do not use. Use "plug-in".

I see at least one other instance to update, but please fix them all.

> * The `kn func emit` command has been added to the functions `kn` plugin. This command allows you to send events to test locally deployed functions.

Reviewer 2: > ! allow verb
Avoid stating that inanimate objects grant abilities to people, as in "the product allows you to...". Whenever possible, use a direct, user-focused alternative...

[id="fixed-issues-1-16-0_{context}"]
== Fixed issues

[id="known-issues-1-16-0_{context}"]
== Known issues

// Add a note about the following to the mTLS docs when they're ready to merge
* If Service Mesh is enabled with mTLS, metrics for Knative Serving are disabled by default because Service Mesh prevents Prometheus from scraping metrics.
+

> If you want to enable Knative Serving metrics for use with Service Mesh and mTLS, you must complete the following steps:

Reviewer 2: It would be a better, kinder user experience to include the commands to edit and apply these YAML files. If you're confident that your target user will be able to locate all of these (especially the Service Mesh control plane object), feel free to ignore this suggestion.

Author: I think it's OK for now since things like the Service Mesh control plane object are documented in more depth elsewhere

.. Specify `prometheus` as the `metrics.backend-destination` in the `observability` spec of the Knative Serving custom resource (CR):
+
[source,yaml]
----
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    observability:
      metrics.backend-destination: "prometheus"
----
+
This step prevents metrics from being disabled by default.

.. Apply the following network policy to allow traffic from the Prometheus namespace:
+
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-from-openshift-monitoring-ns
  namespace: knative-serving
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          name: "openshift-monitoring"
  podSelector: {}
  policyTypes:
  - Ingress
----

.. Modify and reapply the default Service Mesh control plane in the `istio-system` namespace, so that it includes the following spec:
+
[source,yaml]
----
spec:
  proxy:
    networking:
      trafficControl:
        inbound:
          excludedPorts:
          - 8444
----

> * If you deploy Service Mesh CRs with the Istio ingress enabled, you may see the following warning in the `istio-ingressgateway` pod:

Reviewer 2: > !  may verb
Use only to indicate permission. Otherwise, use "can" to indicate ability, or use "might" to indicate possibility. "May" is overused in technical information. In most cases, use "can" or "might" instead.

This one reads like "might." And I see at least one other instance of "may," so please check the rest of the PR and make updates.

+
[source,terminal]
----
2021-05-02T12:56:17.700398Z warning envoy config [external/envoy/source/common/config/grpc_subscription_impl.cc:101] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Error adding/updating listener(s) 0.0.0.0_8081: duplicate listener 0.0.0.0_8081 found
----
+
Your Knative services may also not be accessible.
+
You can use the following workaround to fix this issue by recreating the `knative-local-gateway` service:

.. Delete the existing `knative-local-gateway` service in the `istio-system` namespace:
+
[source,terminal]
----
$ oc delete services -n istio-system knative-local-gateway
----

> .. Create and apply a new `knative-local-gateway` service that contains the following YAML:

Reviewer 2: This is one of my pet peeves: "create ... a new" is redundant. I suggest removing "new."

+
[source,yaml]
----
apiVersion: v1
kind: Service
metadata:
 name: knative-local-gateway
 namespace: istio-system
 labels:
   experimental.istio.io/disable-gateway-port-translation: "true"
spec:
 type: ClusterIP
 selector:
   istio: ingressgateway
 ports:
   - name: http2
     port: 80
     targetPort: 8081
----

> * If you have 1000 Knative services on a cluster, and then perform a reinstall or upgrade of Knative Serving, there is a delay when you create the first new service after the `KnativeServing` custom resource definition (CRD) becomes Ready.

Reviewer 2: ```suggestion
* If you have 1000 Knative services on a cluster and then perform a reinstall or upgrade of Knative Serving, there is a delay when you create the first new service after the `KnativeServing` custom resource definition (CRD) becomes `Ready`.
```
Double-check the repo guidelines for [backticks](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference), but from context, I think "Ready" needs them.
